### PR TITLE
ECSB-529 ECSB-506 Fix arm hash crash

### PR DESCRIPTION
**Pull request by unpacklo**
On arm 32 bit, IL2CPP builds can crash due to `math.hash` attempting to access `uint` at unaligned addresses. This PR fixes the problem across all platforms by causing all reads to go through a 32 bit little endian read function which reads byte by byte and then constructs the 32 bit value.

I manually checked this fix on a test project with a Samsung Galaxy S10e (SM-G970U).

### CODE REVIEW

**Comment by unpacklo** [comment]
I do have some concerns about a possible performance regression on x64 but I heard through the grapevine that unaligned loads/stores are no longer performance penalized on recent Intel chips. Not sure about AMD. In practice, my Intel Core i7-7820X has no performance difference.

Does anyone know the story on AMD side? Maybe @tgjones knows?

File: src/Unity.Mathematics/math.cs [hunk 1]
```diff
@@ -6728,27 +6740,28 @@ public static unsafe uint hash(void* pBuffer, int numBytes, uint seed = 0)
                     int count = numBytes >> 4;
                     for (int i = 0; i < count; ++i)
                     {
-                        state += *p++ * Prime2;
-                        state = (state << 13) | (state >> 19);
+                        var data = new uint4(read32_little_endian(p), read32_little_endian(p + 4), read32_little_endian(p + 8), read32_little_endian(p + 12));
```
Comment (unpacklo) [inline src/Unity.Mathematics/math.cs:6743]: This is the fix. Read little endian byte by byte every time instead of assuming `uint`.

File: src/Unity.Mathematics/math.cs [hunk 2]
```diff
@@ -6704,6 +6704,18 @@ public static float4 chgsign(float4 x, float4 y)
             return asfloat(asuint(x) ^ (asuint(y) & 0x80000000));
         }
 
+        /// <summary>
+        /// Read 32 bits of data in little endian format.
+        /// </summary>
+        /// <param name="pBuffer">Memory address to read from.</param>
+        /// <returns>32 bits in little endian format.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe uint read32_little_endian(void* pBuffer)
```
Comment (unpacklo) [inline src/Unity.Mathematics/math.cs:6713]: private so we don't enlarge API surface area.

Comment (unpacklo) [reply src/Unity.Mathematics/math.cs:6713]: Would love for this to be a burst intrinsic. Does it already exist?

Comment (unpacklo) [reply src/Unity.Mathematics/math.cs:6713]: Oh... I can't depend on burst. Oops.

Comment (joncham) [inline src/Unity.Mathematics/math.cs:6713]: Once available on all Unity versions where this is needed, we should probably use [Unsafe.ReadUnaligned](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.readunaligned?view=net-7.0#system-runtime-compilerservices-unsafe-readunaligned-1(system-void*))

**Comment by unpacklo** [comment]
@cdwfs take a gander at this glorious assembly diff https://gist.github.com/unpacklo/864df87d23bdbdee242d3212123cf4d1

There are no appreciable code diffs here. There are a few rearrangements of some code lines and debug info but that seems to be coming from my switch to calling `rol` instead of the handwritten rotate that was in the hash code before.

File: src/Unity.Mathematics.PerformanceTests/TestHash.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace Unity.Mathematics.PerformanceTests
+{
+    [BurstCompile]
+    partial class TestHash
+    {
+        static readonly long kBufferBytes = 1024 * 1024;
+
+        [Test, Performance]
+        public static unsafe void HashAligned()
```
Comment (kevinmv) [inline src/Unity.Mathematics.PerformanceTests/TestHash.cs:16]: Minor: but since these tests don't assert anything, it would be nice if they were named something like `HashAligned_ShouldNotCrash()` so it's clear that you didn't miss adding an assert and that just running is good enough (I mean, _most_ code should not crash as a minimum but here it's the only goal since no other qualities of the hash are being validated)

File: src/Unity.Mathematics.PerformanceTests/TestHash.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            Debug.Log(Hash(false));
+        }
+
+        [Test, Performance]
+        public static unsafe void HashUnaligned()
+        {
+            Debug.Log(Hash(true));
```
Comment (kevinmv) [inline src/Unity.Mathematics.PerformanceTests/TestHash.cs:24]: I suspect the debug log is to ensure Burst doesn't optimize out the code. Might be good to comment that so it doesn't get removed since some tests check for clean logs which this would potentially break if mathematics tests were ever moved into the same project as another manifest.json `testable`

File: Unity.Mathematics.TestProject/Assets/Scenes/HashAligned.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using Unity.Burst;
+using System;
+using Unity.Collections.LowLevel.Unsafe;
+using Unity.Mathematics;
+
+[BurstCompile]
```
Comment (kevinmv) [inline Unity.Mathematics.TestProject/Assets/Scenes/HashAligned.cs:9]: Any reason why a new monobehaviour needs to be added and added to the sample scene? I'm not familiar with math's test setup so maybe this is normal, but I'd figure validation of hash correctness on players would be handled by the standalone testing for the `[Test]`s you added

**Review by kevinmv — APPROVED** [review]
Looks good. Added some minor comments and questions but nothing blocking

File: Unity.Mathematics.TestProject/Assets/Scenes/HashAligned.cs (hunk 5 above)
Comment (unpacklo) [reply Unity.Mathematics.TestProject/Assets/Scenes/HashAligned.cs:9]: It was to test with a player build. The mathematics CI actually doesn't build any players so we have a huge hole in testing there. How do I build all the tests to run in a player?

Comment (kevinmv) [reply Unity.Mathematics.TestProject/Assets/Scenes/HashAligned.cs:9]: The test asmdef is currently specified to only run in editor:
```
    "includePlatforms": [
        "Editor"
    ],
```
https://github.com/Unity-Technologies/Unity.Mathematics/blob/master/src/Tests/Unity.Mathematics.Tests.asmdef#L10

If you change it to not specify the "Editor" paltform it will default to all platforms. e.g.

```
{
    "name": "Unity.Mathematics.Tests",
    "rootNamespace": "",
    "references": [
        "Unity.Mathematics",
        "UnityEngine.TestRunner",
        "UnityEditor.TestRunner"
    ],
    "includePlatforms": [],
    "excludePlatforms": [],
    "allowUnsafeCode": true,
    "overrideReferences": true,
    "precompiledReferences": [
        "nunit.framework.dll"
    ],
    "autoReferenced": false,
    "defineConstraints": [
        "UNITY_INCLUDE_TESTS"
    ],
    "versionDefines": [],
    "noEngineReferences": false
}
```

**Comment by Over17** [comment]
> @cdwfs take a gander at this glorious assembly diff https://gist.github.com/unpacklo/864df87d23bdbdee242d3212123cf4d1
> 
> There are no appreciable code diffs here. There are a few rearrangements of some code lines and debug info but that seems to be coming from my switch to calling `rol` instead of the handwritten rotate that was in the hash code before.

The `rol` 's where in before and after - I can only see instructions rearranged

**Comment by Over17** [comment]
@unpacklo you checked on a Samsung with an Armv7 build, right? just checking to be sure.

**Comment by Over17** [comment]
@unpacklo could you share the arm assembly diff? most interested in arm64, so that it does not regress.

**Comment by unpacklo** [comment]
I spent the last few days debugging the il2cpp debug and release builds of this crash and as a result, I have a clear understanding of what's going on.

4da3001732c904b4773d7d06c4315c1f

First off, this is the android armv7a neon32 build crashing on device:

image

Notice the `ldm r0, {r3, r4, r5, r7}` instruction which is an aligned load. However, `r0` is in fact an unaligned address:

image

I mean, that's cool and all so at least that part is confirmed. However, I noticed during my testing that il2cpp debug builds were running just fine. Time to find out.

The debug disassembly looks a bit different:

image

Here, we are stopped right before the load of interest `vld1.32 {d16, d17} [r0]` and `r0` has the unaligned pointer `0xc71a12d9` (I only have the var window, not the `register read` from lldb but rest assured, r0 is unaligned):

image

I'm a bit confused at this point because I'm vaguely familiar with arm and I thought this was an instruction that required alignment, but somehow this seems to be fine. Well, the almighty Google brought me to this insightful stackoverflow:

image

Screen_Shot_2018-10-25_at_11 02 15_AM

Ok, I have to find out more...

image

image

image

Screen_Shot_2018-10-25_at_11 02 15_AM-1

So basically, if you've got `SCTLR.A == 0` and you use `vld1.32` without any alignment hint, you have an unaligned load which is why the debug build was not crashing.

I'm still working on a fix for this, more info to come tomorrow.

**Comment by Over17** [comment]
Megacool stuff!

Some additional details:

- release build uses `LDM`, and the page you linked says it checks alignment for `word` and signals a fault if failed in ANY mode of the `A` (alignment) bit of `SCTLR`
- debug build uses `VLD1` and it has its own misalignment behavior (as you pointed out above)
- I don't actually know if a debug build sets `A` to 0 (if it even has the privileges to do so)

I guess the fix should not contain any alignment specificators - the function must work with any input, right?
So in an unaligned case, processing some bytes until you're at a good address could be a way to go.
Important is to not put a penalty on Arm64. Armv7 is on its way to sunset, while Arm64 is to stay with us for many decades.

Alignment is an important corporate exercise, isn't it!

**Review by kevinmv — APPROVED** [review]
Thank you for cleaning up the tests and adding the player testing to confirm everything is working well and ensuring code generation is still optimal for arm64

**Comment by unpacklo** [comment]
> > Important is to not put a penalty on Arm64. Armv7 is on its way to sunset, while Arm64 is to stay with us for many decades.
> 
> I think my last commit will achieve this: [7fb7c93](https://github.com/Unity-Technologies/Unity.Mathematics/commit/7fb7c936e9ff1836c5b3d1d9b4a1bb9343c028cd)
> 
> Basically, I make the code go through the slow path only for 32 bit arm. All other platforms continue to use the original code as before which assumes the platform supports unaligned loads.
> 
> There is a bit of a wrinkle here in that we are relying on the fact that AArch64 _allows_ for unaligned loads/stores to normal memory but still requires alignment for any pointer to device memory. If the user somehow got a pointer to device memory and called this function, we will still crash but I think I can consider this an edge case and move on until someone tells me they're hashing device memory.

I think I accidentally edited Yury's comment. In any case, I'm trying to clean up CI and verifying on the customer bug report project before checking in.

**Comment by unpacklo** [comment]
So, to finally close the loop, I was wondering why arm 64 bit does not crash and the reason seems to be this:

image

So AArch64 says that `Normal` memory can support unaligned accesses. I'm sure there's a penalty or something regarding that and could be optimized in the future, but that's a task for future me. It also mentions `SCTLR_ELx.A` where the `x` corresponds to the processor privilege level.

The key distinction here is `Normal` memory and `Device` memory. `Normal` memory is described here:

image

`Device` memory is described here:

image

The key distinction is that `Device` memory can have side effects from simply accessing the memory contents. I think it's safe to say that no one would ever want to be hashing `Device` memory so I am comfortable with assuming that no one would ever send the hash function a `Device` memory pointer.

Finally, the value of `SCTLR_ELx.A`. What is the value set to and how does it get set? Well, these are privileged registers that get set by the OS. I've scrubbed around and I can't quite nail down what android does but empirically, they're doing EL0.A = 0 so unaligned accesses are valid.

**Comment by unpacklo** [comment]
I finally got my hands on the customer bug report test project and I've reproduced the issue. I've verified that this PR fixes the crash on arm32 in the customer test project.

**Comment by Over17** [comment]
Great stuff, thanks Dale!

> Finally, the value of SCTLR_ELx.A. What is the value set to and how does it get set? Well, these are privileged registers that get set by the OS.

You are correct, those are priviliged registers so setting them is not possible from the user space, unless I'm mistaken. We can consider this irrelevant.

RE: the code, I think the slow path will trigger for x86 (intel 32 bit) Android too, which Unity supports only for Chromebooks. I think it's fine though, x86 ABI is even unsupported in latest NDKs, so the share of affected devices is super low.
